core/src/main/java/io/micronaut/core/io/buffer/ByteArrayByteBuffer.java

```diff
@@ -0,0 +1,236 @@
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.core.io.buffer;
+
+import io.micronaut.core.annotation.Experimental;
+import io.micronaut.core.annotation.Internal;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+
+/**
+ * A {@link ByteBuffer} implementation that is backed by a byte array.
+ *
+ * @since 4.7
+ */
+@Internal
+@Experimental
+public class ByteArrayByteBuffer implements ByteBuffer<byte[]> {
+
+    private final byte[] underlyingBytes;
+    private int readerIndex;
+    private int writerIndex;
+
+    /**
+     * Construct a new {@link ByteArrayByteBuffer} for the given bytes.
+     *
+     * @param underlyingBytes the bytes to wrap
+     */
+    public ByteArrayByteBuffer(byte[] underlyingBytes) {
+        this(underlyingBytes, underlyingBytes.length);
+    }
+
+    /**
+     * Construct a new {@link ByteArrayByteBuffer} for the given bytes and capacity.
+     * If capacity is greater than the length of the underlyingBytes, extra bytes will be zeroed out.
+     * If capacity is less than the length of the underlyingBytes, the underlyingBytes will be truncated.
+     *
+     * @param underlyingBytes the bytes to wrap
+     * @param capacity        the capacity of the buffer
+     */
+    public ByteArrayByteBuffer(byte[] underlyingBytes, int capacity) {
+        if (capacity < underlyingBytes.length) {
+            this.underlyingBytes = Arrays.copyOf(underlyingBytes, capacity);
+        } else if (capacity > underlyingBytes.length) {
+            this.underlyingBytes = new byte[capacity];
+            System.arraycopy(underlyingBytes, 0, this.underlyingBytes, 0, underlyingBytes.length);
+        } else {
+            this.underlyingBytes = underlyingBytes;
+        }
+    }
+
+    @Override
+    public byte[] asNativeBuffer() {
+        return underlyingBytes;
+    }
+
+    @Override
+    public int readableBytes() {
+        return underlyingBytes.length - readerIndex;
+    }
+
+    @Override
+    public int writableBytes() {
+        return underlyingBytes.length - writerIndex;
+    }
+
+    @Override
+    public int maxCapacity() {
+        return underlyingBytes.length;
+    }
+
+    @Override
+    public ByteArrayByteBuffer capacity(int capacity) {
+        return new ByteArrayByteBuffer(underlyingBytes, capacity);
+    }
+
+    @Override
+    public int readerIndex() {
+        return readerIndex;
+    }
+
+    @Override
+    public ByteArrayByteBuffer readerIndex(int readPosition) {
+        this.readerIndex = Math.min(readPosition, underlyingBytes.length - 1);
```

Comment:
```suggestion
        this.readerIndex = Math.min(readPosition, underlyingBytes.length);
```

Reply:
Are you sure? I just copied it from Servlet

Reply:
Yes, otherwise you can never set readerIndex so that readableBytes == 0